src/units/animation.cpp

```diff
@@ -1329,6 +1373,196 @@ void unit_animator::add_animation(unit_const_ptr animated_unit
 }
 
+bool unit_animator::move_unit_fake(int& index_movement_anim)
+{
+	const std::string tag = "move_unit_fake";
+	game_events::wml_action::map m = game_events::wml_action::registry();// m[tag] is call wml func by tag.
+	game_display* screen_p = game_display::get_singleton();
+	display* disp = display::get_singleton();
+	bool result;
+	bool coherent_moving = false;
+	map_location u_initial_loc;
+	while (index_movement_anim < animated_units_.size() && animated_units_[index_movement_anim].is_movement)
+	{
+		if(index_movement_anim + 1 < animated_units_.size()
+			&& animated_units_[index_movement_anim].serial_no == animated_units_[index_movement_anim + 1].serial_no) {
+			return true;
+		}
+		auto& anim = animated_units_[index_movement_anim];
+		auto* last_anim_p = index_movement_anim > 0 ? &animated_units_[index_movement_anim - 1] : nullptr;
+		// only when start location and animated unit are the same with what in last animation, should the next movement animation run in coherence.
+		coherent_moving = last_anim_p && anim.coherence && last_anim_p->is_movement
+			&& last_anim_p->move_up == anim.move_up && last_anim_p->src == anim.src;
+		unit_ptr& m_up = anim.move_up;
+		unit& m_u = *m_up;
+		map_location& m_src = coherent_moving ? last_anim_p->move_dst : anim.src;
+		map_location& m_dst = anim.move_dst;
+		// when running the animation, loc of the real unit could be different from m_src, and the former is needed
+		u_initial_loc = m_u.get_location();
+		m_u.set_hidden(true);
+		// temporarily changes the unit location to make space.
+		if (coherent_moving){
+			m_u.set_location(last_anim_p->move_dst);
+		} else {
+			m_u.set_location(map_location::null_location());
+		}
+		disp->invalidate(u_initial_loc);
+		m_dst = find_vacant_tile(m_dst, pathfind::VACANT_ANY, m_up.get());
+		config cfg;
+		cfg["type"] = m_u.type().id();
+		cfg["gender"] = m_u.gender();
+		cfg["variation"] = m_u.variation();
+		cfg["side"] = m_u.side();
+		cfg["x"] = std::to_string(m_src.wml_x()) + "," + std::to_string(m_dst.wml_x());
+		cfg["y"] = std::to_string(m_src.wml_y()) + "," + std::to_string(m_dst.wml_y());
+		cfg["force_scroll"] = true;
+		vconfig vcfg = vconfig(cfg);
+		m[tag]({"","",map_location(),map_location(),config()},vcfg);
+		result = screen_p->maybe_rebuild();
+		if (!result){
+			screen_p->invalidate_all();
+		}
+		m_u.set_location(u_initial_loc);
+		index_movement_anim++;
+	}
+	disp->update();
+	return index_movement_anim < animated_units_.size() && animated_units_[index_movement_anim].is_movement;
+}
+
+bool unit_animator::move_units_fake(int& index, const int& size)
+{
+	if(size < 1) {
+		if(index < animated_units_.size()) {
+			return animated_units_[index].is_movement;
+		} else {
+			return false;
+		}
+	}
+	const int start = index;
+	int temp_index = index;
+	const std::string tag = "move_units_fake";
+	game_events::wml_action::map m = game_events::wml_action::registry(); // m[tag] is call wml func by tag.
```

Comment:
Woah, this should be a const reference, not a copy! We don't want to duplicate the whole wml action registry. Besides, if it's only used once, you can just dispense with the variable and use `game_events::wml_action::registry()` directly.